File: 06-adapter/06-adapter.py
from abc import ABC, abstractmethod

class USBC(ABC):
    @abstractmethod
    def connect(self):
        raise NotImplementedError("Subclass must implement connect()")

class Phone(USBC):
    def connect(self):
        return "phone charging"

class Headphones:
    def plug(self):
        return "adding connector"

class HeadphonesUSBC(USBC):
    def __init__(self, headphones: Headphones):
        self.headphones = headphones

    def connect(self):
        return self.headphones.plug() + " and headphones charging."


headphones = Headphones()
headphones_adapter = HeadphonesUSBC(headphones)
print(headphones_adapter.connect())

File: 06-adapter/06-adapter.java
interface EUSocket {
    public String connect();
}

class TV implements EUSocket {
    @Override
    public String connect() {
        return "tv connected to eu socket";
    }
}

// Adaptee
class Razor {
    public String plug() {
        return "Adding UK adapter";
    }
}

class RazorWithAdapter implements EUSocket {
    private Razor razor;

    public RazorWithAdapter(Razor razor) {
        this.razor = razor;
    }

    @Override
    public String connect() {
        return this.razor.plug() + " and razor connected to eu socket";
    }
}

public class Main {
    public static void main(String []args) {
        Razor r = new Razor();
        RazorWithAdapter ra = new RazorWithAdapter(r);
        System.out.println(ra.connect());
    }
}
